lib/core.js

```diff
@@ -35,12 +35,12 @@ function Promise(fn) {
       deferreds.push(deferred)
       return
     }
+    var cb = state ? deferred.onFulfilled : deferred.onRejected
+    if (cb === null) {
+      (state ? deferred.resolve : deferred.reject)(value)
+      return
+    }
     asap(function() {
-      var cb = state ? deferred.onFulfilled : deferred.onRejected
-      if (cb === null) {
-        (state ? deferred.resolve : deferred.reject)(value)
-        return
-      }
```

Comment:
imo, order depending on value + callback type is definitely not good.
`.then()` should behave as `.then(id, rethrow)`
`.then(null, f)` should behave as `.then(id, f)`
`.then(f, null)` should behave as `.then(f, rethrow)`